File: socket/v3-chat/RUBY/chat_handler.rb
require 'thread'
require 'time'

module ChatHandler
  MAX_CLIENTS = 64

  @@clients = []
  @@mutex = Mutex.new

  class << self
    def timestamp
      Time.now.strftime("%Y-%m-%d %H:%M:%S")
    end

    def broadcast(message)
      sockets = []

      @@mutex.synchronize do
        sockets = @@clients.map { |c| c[:socket] }
      end

      sockets.each do |sock|
        begin
          sock.write("#{message}\n")
        rescue
        end
      end
    end

    def online_list
      @@mutex.synchronize do
        @@clients.map { |c| c[:name] }.join(" ")
      end
    end

    def register_client(sock, name)
      @@mutex.synchronize do
        return false if @@clients.length >= MAX_CLIENTS
        return false if @@clients.any? { |c| c[:name] == name }

        @@clients << { socket: sock, name: name }
        true
      end
    end

    def remove_client(sock)
      removed = nil

      @@mutex.synchronize do
        idx = @@clients.index { |c| c[:socket] == sock }
        if idx
          removed = @@clients[idx][:name]
          @@clients.delete_at(idx)
        end
      end

      removed
    end

    def handle_client(conn)
      name = nil

      begin
        name = conn.gets
        if name.nil?
          conn.close
          return
        end

        name = name.strip
        if name.empty?
          conn.write("Servidor: nome de usuario vazio.\n")
          conn.close
          return
        end

        unless register_client(conn, name)
          conn.write("Servidor: nome ja em uso ou sala cheia.\n")
          conn.close
          return
        end

        conn.write("Servidor: bem-vindo, #{name}!\n")
        current_online = online_list
        conn.write("Servidor: usuarios online agora -> #{current_online.empty? ? '(ninguem)' : current_online}\n")

        puts "[#{timestamp}] #{name} entrou no chat."
        broadcast("Servidor: [#{name}] entrou no chat.")

        loop do
          line = conn.gets
          break if line.nil?

          msg = line.gsub(/\r?\n/, "")
          next if msg.empty?
          break if msg == "/quit"

          message = "[#{timestamp}] #{name}: #{msg}"
          puts message
          broadcast(message)
        end
      rescue => e
        warn "[-] Erro com cliente: #{e.message}"
      ensure
        removed = remove_client(conn)
        if removed
          puts "[#{timestamp}] #{removed} saiu do chat."
          broadcast("Servidor: [#{removed}] saiu do chat.")
        end
        conn.close rescue nil
      end
    end
  end
end

File: socket/v3-chat/PYTHON/chat_handler.py
import threading
import time

MAX_CLIENTS = 64

clients = []
clients_lock = threading.Lock()

def timestamp():
    return time.strftime("%Y-%m-%d %H:%M:%S")

def send_line(sock, text):
    sock.sendall((text + "\n").encode("utf-8", errors="replace"))

def broadcast(message):
    with clients_lock:
        snapshot = [c["socket"] for c in clients]

    for sock in snapshot:
        try:
            send_line(sock, message)
        except Exception:
            pass

def online_list():
    with clients_lock:
        return " ".join(c["name"] for c in clients)

def register_client(sock, name):
    with clients_lock:
        if len(clients) >= MAX_CLIENTS:
            return False
        if any(c["name"] == name for c in clients):
            return False
        clients.append({"socket": sock, "name": name})
        return True

def remove_client(sock):
    with clients_lock:
        for i, c in enumerate(clients):
            if c["socket"] == sock:
                removed = c["name"]
                clients.pop(i)
                return removed
    return None

def handle_client(client_socket, addr):
    print(f"[+] Cliente conectado: {addr}")
    name = None

    try:
        with client_socket:
            reader = client_socket.makefile("r", encoding="utf-8", newline="\n")

            first_line = reader.readline()
            if not first_line:
                return

            name = first_line.strip()

            if not name:
                send_line(client_socket, "Servidor: nome de usuario vazio.")
                return

            if not register_client(client_socket, name):
                send_line(client_socket, "Servidor: nome ja em uso ou sala cheia.")
                return

            send_line(client_socket, f"Servidor: bem-vindo, {name}!")
            current_online = online_list()
            send_line(
                client_socket,
                f"Servidor: usuarios online agora -> {current_online if current_online else '(ninguem)'}"
            )

            print(f"[{timestamp()}] {name} entrou no chat.")
            broadcast(f"Servidor: [{name}] entrou no chat.")

            while True:
                line = reader.readline()
                if not line:
                    break

                msg = line.rstrip("\r\n")
                if not msg:
                    continue

                if msg == "/quit":
                    break

                message = f"[{timestamp()}] {name}: {msg}"
                print(message)
                broadcast(message)

            reader.close()

    except Exception as e:
        print(f"[-] Erro na conexao com {addr}: {e}")

    finally:
        removed_name = remove_client(client_socket)
        if removed_name:
            print(f"[{timestamp()}] {removed_name} saiu do chat.")
            broadcast(f"Servidor: [{removed_name}] saiu do chat.")
        print(f"[-] Cliente {addr} desconectou.")